lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+    ],
+    reg_covar: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: """
+      The non-negative number that is added to each element of the diagonal
+      of the covariance matrix to ensure it is positive.
+      """
+    ],
+    key: [
+      type: {:custom, Scholar.Options, :key, []},
+      doc: """
+      Used for random number generation in parameter initialization.
+      If the key is not provided, it is set to `Nx.Random.key(System.system_time())`.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a Gaussian Mixture Model for sample inputs `x`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:weights` - The fractions of data sampled from each Gaussian, respectively.
+
+    * `:means` - Means of the Gaussian components.
+
+    * `:covariances` - Covariance matrices of the Gaussian components.
+
+    * `:precisions_cholesky` - Cholesky decomposition of the precision matrices
+  	(inverses of covariances). This is useful for the model inference.
+
+  ## Examples
+
+      iex> key = Nx.Random.key(12)
+      iex> x = Nx.tensor([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
+      iex> Scholar.Cluster.GaussianMixture.fit(x, num_gaussians: 2, key: key).means
+      Nx.tensor(
+        [
+          [1.0, 2.0],
+          [10.0, 2.0]
+        ]
+      )
+  """
+  deftransform fit(x, opts \\ []) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            """
+            expected input tensor to have shape {n_samples, n_features}, \
+            got tensor with shape: #{inspect(Nx.shape(x))}\
+            """
+    end
+
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    num_samples = Nx.axis_size(x, 0)
+
+    unless opts[:num_gaussians] <= num_samples do
+      raise ArgumentError,
+            """
+            invalid value for :num_gaussians option: \
+            expected positive integer between 1 and #{inspect(num_samples)}, \
+            got: #{inspect(opts[:num_gaussians])}\
+            """
+    end
+
+    key = Keyword.get_lazy(opts, :key, fn -> Nx.Random.key(System.system_time()) end)
+    fit_n(x, key, opts)
+  end
+
+  defnp fit_n(x, key, opts) do
+    x = to_float(x)
+    num_gaussians = opts[:num_gaussians]
+    num_runs = opts[:num_runs]
+    max_iter = opts[:max_iter]
+    tol = opts[:tol]
+    reg_covar = opts[:reg_covar]
+
+    {best_params, _} =
+      while {
+              best_params =
+                init_params(key, x, num_gaussians: num_gaussians, reg_covar: reg_covar),
+              {
+                run = 1,
+                max_lower_bound = Nx.Constants.neg_infinity(),
+                x,
+                key,
+                max_iter,
+                tol,
+                reg_covar
+              }
+            },
+            run <= num_runs do
```

Comment:
Since there's no early stopping, we might be able to apply vectorization to get rid of this outer while loop

Reply:
I think that in the case of k-means, vectorization is more beneficial. You can simply reduce `num_runs` if the memory is the problem. But you cannot apply vectorization programmatically if you use two nested whiles. Nowadays, computation time is much more important than space limitation. You might benchmark two versions on the example that can be loaded into memory and test how much faster the vectorized version is. If the change is insignificant for GMM, we should stay with a memory-efficient solution.

Reply:
And it's always easier to change vectorized version into 'while', so if people had a memory problem on k-means, then we would use a memory-efficient solution.

Reply:
And in Sci-kit they cannot run k-means on GPU, so they have no clear advantage of implementing vectorization which is not obvious to understand at first glance, harder to implement (because of broadcasting), memory consuming etc.